### PR TITLE
1466015: Warn of deprecation of command line options in next release

### PR DESCRIPTION
**Pull request by wottop**

### CODE REVIEW

**Comment by wottop** [comment]
[1466015 - [RFE] Add warning that we will remove backend CLI configuration options in a future release](https://trello.com/c/jQwPIGeK/514-ions)

File: virt-who-config.5 [hunk 1]
```diff
@@ -23,7 +23,7 @@ The following are options that can be specified in the \fBglobal\fR section:
 how often to check connected hypervisors for changes (seconds). Also affects how often a mapping is reported.
 .TP
 \fBreporter_id\fR
-The id of this virt-who instance, reported with all mappings.
```
Comment (cnsnyder) [inline virt-who-config.5]: These options will still be allowable in the configuration files.
The man pages changes necessary are in ./virt-who.8

Please update that man page accordingly.

Thank you!!

File: virtwho/parser.py [hunk 2]
```diff
@@ -456,9 +457,14 @@ def parse_options():
...
     :return: Tuple of logger and options
     """
+    # These options are deprecated
+    DEPRECATED_OPTIONS = ['log_per_config', 'log_dir', 'log_file', 'reporter_id']
 
     # Read command line arguments first
     cli_options, defaults = parse_cli_arguments()
+    for option in DEPRECATED_OPTIONS:
+        if option in cli_options and not cli_options[option] == defaults[option]:
+            print("The option --%s is deprecated and will be removed in the next release." % option)
```
Comment (cnsnyder) [inline virtwho/parser.py]: I would add something to the effect of "please see 'man virt-who-config' for details on adding a configuration section"

File: virtwho/parser.py [hunk 3]
```diff
@@ -310,15 +310,15 @@ def parse_cli_arguments():
...
-                        help="The absolute path of the directory to log to. (Default '%s')" % log.DEFAULT_LOG_DIR)
+                        help="[Deprecated] The absolute path of the directory to log to. (Default '%s')" % log.DEFAULT_LOG_DIR)
     parser.add_argument("-f", "--log-file", action="store", dest="log_file", default=log.DEFAULT_LOG_FILE,
-                        help="The file name to write logs to. (Default '%s')" % log.DEFAULT_LOG_FILE)
+                        help="[Deprecated] The file name to write logs to. (Default '%s')" % log.DEFAULT_LOG_FILE)
     parser.add_argument("-r", "--reporter-id", action="store", dest="reporter_id", default=NotSetSentinel(),
-                        help="Label host/guest associations obtained by this instance of virt-who with the provided id.")
+                        help="[Deprecated] Label host/guest associations obtained by this instance of virt-who with the provided id.")
 
     virt_group = parser.add_argument_group(
```
Comment (cnsnyder) [inline virtwho/parser.py:323]: All options in the virt_group as well as satellite_group will be deprecated. (as we are moving to configuration file only).

**Review by cnsnyder — CHANGES_REQUESTED** [review]
Please see the comments above. A few more options need to be listed as deprecated.

Thank you!

**Review by cnsnyder — APPROVED** [review]
Nevermind, there are still options that need to be listed as deprecated.

All options that allow command line configuration of virt backends are to be deprecated.
Any that match the below (glob-like) expression:
```--*-{username,env,owner,server,password}```
Should be deprecated.

**Review by cnsnyder — CHANGES_REQUESTED** [review]
Nevermind, there are still options that need to be listed as deprecated.

All options that allow command line configuration of virt backends are to be deprecated.
Any that match the below (glob-like) expression:
```--*-{username,env,owner,server,password}```
Should be deprecated.

File: virtwho/parser.py [hunk 4]
```diff
@@ -456,9 +457,17 @@ def parse_options():
     This function parses all options from command line and environment variables
     :return: Tuple of logger and options
     """
+    # These options are deprecated
+    DEPRECATED_OPTIONS = ['log_per_config', 'log_dir', 'log_file', 'reporter_id',
```
Comment (cnsnyder) [inline virtwho/parser.py]: Please update this to include all deprecated options above.

**Review by cnsnyder — CHANGES_REQUESTED** [review]
Please update this to include all deprecated options per prior changes. Thank you.

File: virtwho/parser.py [hunk 5]
```diff
@@ -456,6 +457,14 @@ def parse_options():
...
+                          'libvirt-owner', 'libvirt-env', 'libvirt-server', 'libvirt-username', 'libvirt-password',
+                          'esx-owner', 'esx-env', 'esx-server', 'esx-username', 'esx-password',
+                          'rhevm-owner', 'rhevm-env', 'rhevm-server', 'rhevm-username', 'rhevm-password',
+                          'hyperv-owner', 'hyperv-env', 'hyperv-server', 'hyperv-username', 'hyperv-password',
+                          'xen-owner', 'xen-env', 'xen-server', 'xen-username', 'xen-password',
+                          'satellite-server', 'satellite-username', 'satellite-password']
```
Comment (cnsnyder) [inline virtwho/parser.py]: I think we'll need to add 'libvirt', 'vdsm', 'esx', 'xen', 'rhevm', 'hyperv'.
These also should have their help text updated to include '[Deprecated] '.

**Review by cnsnyder — APPROVED** [review]
LGTM!

Thank you for the contribution!

**Review by cnsnyder — CHANGES_REQUESTED** [review]
Tests do not pass.
It is because of the extra 'i' on line 327 (it appears).

After the removal of this there is one failing test, which does not fail in master.

It appears the failure is due to one of the DEPRECATED_OPTIONS not being in the  cli_options dict. I think this is because the options end up in that dict with all '-' changed to '_'.

Sorry for the back and forth...

Cheers

**Review by cnsnyder — APPROVED** [review]
Looks good to me. Merging!

**Pull request merged**